source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+  // Photons are being sorted every pseudo time step, thus no need to sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  cells_local_neighbor_list.clear();
+  cells_ghost_neighbor_list.clear();
+
+  // Update cell neighbors
+  find_cell_neighbors<dim, true>(triangulation,
+                                 cells_local_neighbor_list,
+                                 cells_ghost_neighbor_list);
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::print_insertion_info(const unsigned int &inserted_photon,
+                                            const ConditionalOStream &pcout)
+{
+  std::stringstream ss;
+
+  ss << inserted_photon << " photons "
+     << " were inserted.";
+
+  announce_string(pcout, ss.str(), '*');
+}
+
+
+template <int dim>
+void
+RayTracingSolver<dim>::insert_particles_and_photons()
+{
+  // Insert particles using the insertion object.
+  particle_insertion_object->insert(particle_handler,
+                                    triangulation,
+                                    dem_parameters);
+
+  // A vector which contains all the insertion point of every photon.
+  std::vector<Point<dim>> insertion_points_on_proc;
+
+  // A vector of vectors, which contains all the properties of all particles
+  // about to get inserted.
+  // Each photon has as its properties its insertion location. This way,
+  // when many intersection points will be found during a pseudo timestep,
+  // its properties will be used to identify the closest point from its
+  // insertion insertion point.
+  std::vector<std::vector<double>> photon_properties;
+
+  unsigned int max_n_photon_first_dir =
+    parameters.ray_tracing_info.number_of_photon_first_direction;
+
+  unsigned int max_n_photon_second_dir =
+    parameters.ray_tracing_info.number_of_photon_second_direction;
+
+  // Processor 0 will be the only one inserting photon
+  const unsigned int n_particles_to_insert_this_proc =
+    this_mpi_process == 0 ? max_n_photon_first_dir * max_n_photon_second_dir :
+                            0;
+
+  insertion_points_on_proc.reserve(n_particles_to_insert_this_proc);
+
+  photon_properties.reserve(n_particles_to_insert_this_proc);
+
+  // Create the photon insertion location
+  if (this_mpi_process == 0)
+    {
+      // Create variable for readability
+      Point<dim> starting_insertion_point =
+        parameters.ray_tracing_info.starting_point;
+      Tensor<1, dim> first_dir  = parameters.ray_tracing_info.first_direction_unit;
+      Tensor<1, dim> second_dir = parameters.ray_tracing_info.second_direction_unit;
+
+      const double step_first_dir =
+        parameters.ray_tracing_info.step_between_photons_first_direction;
+      const double step_second_dir =
+        parameters.ray_tracing_info.step_between_photons_second_direction;
+
+      // Generation the photon insertion points.
+      Point<dim> temp_point{};
```

Comment:
There is a default constructor for points, no need for the {}
```suggestion
Point<dim> temp_point;